[SPARK-19834][SQL] csv escape of quote escape (prev)

ep1804 opened the pull request:
### What changes were proposed in this pull request?

Escape of escape should be considered when using the UniVocity csv encoding/decoding library.

I added lines for this.

ref: http://stackoverflow.com/questions/42607208/spark-csv-error-when-reading-backslash-and-quotation-mark


### How was this patch tested?

unit test added

Commit:
SPARK-19384 csv escape of escape

AmplabJenkins commented:
Can one of the admins verify this patch?

ep1804 commented:
cc @HyukjinKwon

HyukjinKwon reviewed (commented):
Thanks for cc'ing me @ep1804. I left my opinion on the codes. Just to make sure, it is not related with actual file encoding/decoding (e.g., encoding option in CSV) and this PR proposes to escape a quote escape, for example, if `espace` is `\` (as `\`), then, `"\\"a"` (raw text) is read `\"a` (token)?

It would be nicer if the PR description has a summary of https://github.com/uniVocity/univocity-parsers/blob/master/README.md#escaping-quote-escape-characters or https://github.com/uniVocity/univocity-parsers/blob/e34a149077147de7cafcafbbaf8a4310e6297584/src/main/java/com/univocity/parsers/csv/CsvFormat.java#L160-L176.

Also, it is a bit confusing to me what this PR proposes. Could you maybe provide input/output and before/after if it does not sound demanding?

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala:
```diff
@@ -167,6 +168,7 @@ private[csv] class CSVOptions(
     format.setDelimiter(delimiter)
     format.setQuote(quote)
     format.setQuoteEscape(escape)
+    format.setCharToEscapeQuoteEscaping(quote)
```
It seems in https://github.com/uniVocity/univocity-parsers/blob/master/README.md#escaping-quote-escape-characters it sets the same `escape` for both `setQuoteEscape` and `setCharToEscapeQuoteEscaping`and I sounds correct to me. However, it seems different here. Do you mind if I ask the reason?

ep1804 commented:
Thank you for early and detailed response @HyukjinKwon .

1. About the purpose of PR, Yes, it's about using escape-a-quote-escape option. I used the wording 'encoding/decoding' with a general meaning and this might be confusing.

2. The indentation errors and readability problem will be fixed following your comments.

3. About the last question, I think this is a problem, and this is because the uniVocity library doesn't work as expected, and I seek your advice. Here's my experiment:

### Test code

For the experiment, I made an additional option `escapeEscape`.

```scala
      val df1 = spark.sqlContext.createDataFrame(List(
        (1, """AA"BB"""),      // 1 quote char (OK without escapeEscape option)
        (2, """AA\"BB"""),     // 1 escape char and 1 quote char
        (3, """AA\\"BB"""),    // 2 escape char and 1 quote char
        (4, """AA""BB"""),     // 2 quote char (OK without escapeEscape option)
        (5, """AA\"\"BB"""),   // (1 escape char anc 1 quote char) * 2
        (6, """AA\\"\\"BB""")  // (2 escape char and 1 quote char) * 2
      ))

      df1.coalesce(1).write
        .format("csv")
        .option("quote", "\"")
        .option("escape", "\\")
        .option("escapeEscape", "\\")
        .save(csvDir)

      val df2 = spark.read
        .format("csv")
        .option("quote", "\"")
        .option("escape", "\\")
        .option("escapeEscape", "\\")
        .load(csvDir).orderBy($"_c0")

```
### Firstly, I set `escapeEscape` to `\`, as documented by uniVocity.

```
+---+----------+
| _1|        _2|
+---+----------+
|  1|     AA"BB|
|  2|    AA\"BB|
|  3|   AA\\"BB|
|  4|    AA""BB|
|  5|  AA\"\"BB|
|  6|AA\\"\\"BB|
+---+----------+

"AA\"BB"
"AA\\"BB"
"AA\\\"BB"
"AA\"\"BB"
"AA\\"\\\"BB"
"AA\\\"\\\\\"BB"

+---+---------+
|_c0|      _c1|
+---+---------+
|  1|    AA"BB|
|  2| "AA\"BB"|
|  3|   AA\"BB|
|  4|   AA""BB|
|  5|  AA\\"BB|
|  6|AA\"\\"BB|
+---+---------+
```
In the result, the first table is df1, the second table is df2, and strings between them are raw csv file content.

In the result, df1 and df2 are different. IMHO, this will not be a result users expect.


### Secondly, I set `escapeEscape` to `A` to test what happens when the escape character meets one pre-existing one.
```
+---+----------+
| _1|        _2|
+---+----------+
|  1|     AA"BB|
|  2|    AA\"BB|
|  3|   AA\\"BB|
|  4|    AA""BB|
|  5|  AA\"\"BB|
|  6|AA\\"\\"BB|
+---+----------+

"AA\"BB"
"AA\\"BB"
"AA\\\"BB"
"AA\"\"BB"
"AA\\"A\\"BB"
"AA\\\"A\A\\"BB"

+---+--------+
|_c0|     _c1|
+---+--------+
|  1|  "\"BB"|
|  2|    \"BB|
|  3|   \\"BB|
|  4|    \"BB|
|  5|  \"\"BB|
|  6|\\"\\"BB|
+---+--------+
```

Again, df1 and df2 are different. Something bad had happened.

### Thirdly, I tested with `escapeEscape = quote` setting

```
+---+----------+
| _1|        _2|
+---+----------+
|  1|     AA"BB|
|  2|    AA\"BB|
|  3|   AA\\"BB|
|  4|    AA""BB|
|  5|  AA\"\"BB|
|  6|AA\\"\\"BB|
+---+----------+

"AA\"BB"
"AA\\"BB"
"AA\\\"BB"
"AA\"\"BB"
"AA\\""\\"BB"
"AA\\\""\"\\"BB"

+---+----------+
|_c0|       _c1|
+---+----------+
|  1|     AA"BB|
|  2|    AA\"BB|
|  3|   AA\\"BB|
|  4|    AA""BB|
|  5|  AA\"\"BB|
|  6|AA\\"\\"BB|
+---+----------+
```


### Finally, I tested `escapeEscape = quote` case with more examples.

```scala
      val df1 = spark.sqlContext.createDataFrame(List(
        (1, """AA\"\"\"\"\"BB"""),   // (1 escape char anc 1 quote char) * 5
        (2, """AA\\"\\"\\"\\"\\"BB"""),  // (2 escape char and 1 quote char) * 5
        (3, "You are \"beautiful\""),
        (4, "Yes, \\\"inside\"\\")
      ))
```

And the result is OK

```
+---+-------------------+
| _1|                 _2|
+---+-------------------+
|  1|     AA\"\"\"\"\"BB|
|  2|AA\\"\\"\\"\\"\\"BB|
|  3|You are "beautiful"|
|  4|    Yes, \"inside"\|
+---+-------------------+

"AA\\""\\""\\""\\""\\"BB"
"AA\\\""\"\\""\"\\""\"\\""\"\\"BB"
"You are \"beautiful\""
"Yes, "\\"inside\""\"

+---+-------------------+
|_c0|                _c1|
+---+-------------------+
|  1|     AA\"\"\"\"\"BB|
|  2|AA\\"\\"\\"\\"\\"BB|
|  3|You are "beautiful"|
|  4|    Yes, \"inside"\|
+---+-------------------+
```

This is why I used `quote` for `escapeEscape` parameter instead of `escape` character.

Commit:
csv escapeEscape option

Commit:
csv comment fix

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala:
```diff
@@ -88,6 +88,7 @@ private[csv] class CSVOptions(
 
   val quote = getChar("quote", '\"')
   val escape = getChar("escape", '\\')
+  val escapeEscape = getChar("escapeEscape", quote)
```
@ep1804, let's use `'\0'` by default as it is the default for `setCharToEscapeQuoteEscaping` if not set according to https://github.com/uniVocity/univocity-parsers/blob/e34a149077147de7cafcafbbaf8a4310e6297584/src/main/java/com/univocity/parsers/csv/CsvFormat.java#L160-L176 because I am worried of changing the existing behaviour for lower versions. Maybe, it is okay for now if it is configurable.

HyukjinKwon replied:
Also, I would make this as an option for this CSV datasource. I think we should document this in `DataFrameReader`, `DataStreamReader`, `readwriter.py` and `streaming.py`. 

Lastly, how about making this from `escapeEscape ` to `escapeQuoteEscaping` in order to take after the original name.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -465,6 +465,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+    }
+  }
+
+
```
super minor: single newline :).

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -465,6 +465,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+    withTempDir { dir =>
+      val csvDir = new File(dir, "csv").getCanonicalPath
+
+
```
super minor nit: here too, single newline if more commits should be pushed.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -465,6 +465,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
   }
 
+  test("save csv with quote escaping enabled, using char to escape quote-escape") {
+    withTempDir { dir =>
```
Let's use `withTempPath { path =>` directly.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -465,6 +465,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+      val df1Str = df1.collect().map(_.getString(1)).mkString("")
+      val df2Str = df2.collect().map(_.getString(1)).mkString("")
+
+      assert(df1Str == df2Str)
```
Maybe we could use `checkAnswer(df1, df2)`. Does this follow your indention?

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -465,6 +465,44 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+        .option("quote", "\"")
+        .option("escape", "\\")
+        .option("escapeEscape", "\"")
+        .load(csvDir).orderBy($"_c0")
```
Per https://github.com/apache/spark/pull/17177/files#r105307099, I think we would not need this `orderBy`.

HyukjinKwon reviewed (commented):
Thanks for taking my opinion into account. I left some more comments. I think I need one more final look. Per https://github.com/apache/spark/pull/17177#issuecomment-284607257, should we maybe open an issue there and check if these are bugs or expected?

In general, I am worried if this looks not working as expected. I am okay with proceeding this if we know the reasons for the behaviour.

So, I think we should be sure on this ^ first. Let me take another look for this option and input/output soon.

Commit:
escapeQuoteEscaping

Commit:
escapeQuoteEscaping again

ep1804 commented:
Thank you @HyukjinKwon .

I made changes following your comments:
 * `escapeQuoteEscaping` instead of `escapeEscape`
 * defalutl value to `\u0000` (unset)
 * `withTempPath`
 * no `orderBy`
 * `checkAnswer`
 * styles : )

ep1804 commented:
An issue is raised for uniVocity parser: https://github.com/uniVocity/univocity-parsers/issues/143

Commit:
doc for escapeQuoteEscaping and python

ep1804 commented:
Documentation for DataFrameReader, DataFrameWriter, DataStreamReader, readwriter.py and streaming.py are written. Check please.

HyukjinKwon commented on python/pyspark/sql/readwriter.py:
```diff
@@ -693,8 +697,8 @@ def text(self, path, compression=None):
     def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=None,
-            header=None, nullValue=None, escapeQuotes=None, quoteAll=None, dateFormat=None,
-            timestampFormat=None, timeZone=None):
+            escapeQuoteEscaping=None, header=None, nullValue=None, escapeQuotes=None, quoteAll=None,
```
Oh, @ep1804, we should add new option at the end of these arguments because otherwise it breaks existing codes using this API in lower versions that use those options as positional arguments (not keyword arguments).

ep1804 replied:
@HyukjinKwon 

It is fixed.

jbax commented:
Doesn't seem correct to me. All test cases are using broken CSV and trigger the parser handling of unescaped quotes, where it tries to rescue the data and produce something sensible. See my test case here: https://github.com/uniVocity/univocity-parsers/issues/143

ep1804 commented:
@jbax 

They are not CSV inputs. They are RAW TEXT.

Commit:
python add argument fix

Commit:
python add argument fix 2

Commit:
more test data

ep1804 commented:
The bug is fixed in uniVocity.  https://github.com/uniVocity/univocity-parsers/issues/143
This will be considered (on Monday).

HyukjinKwon commented:
Thank you so much both for your efforts and time.

Commit:
csv unescapedQuoteHandling option

Commit:
csv escapeUnquotedValues

Commit:
code done

Commit:
escapeUnquotedValues docs

ep1804 commented:
As mentioned in https://github.com/uniVocity/univocity-parsers/issues/143, for proper handling of escape characters, the uniVocity option `escapeUnquotedValues` is also required.

I added that option setter, unit test, and documentation. As for the documentation, I added documentation for `escapeUnquotedValues` to the methods where `escapeQuotes` is documented.

HyukjinKwon commented:
I think it is good to add as it is also described in univocity library but I am not too sure if it is worth exposing an option that has currently a little bug. Maybe we could close for now and re-open this after we bump it up to 2.4.0.

jbax commented:
FYI 2.4.0 is scheduled to be released tomorrow.

On 14 Mar 2017 9:06 PM, "Hyukjin Kwon" <notifications@github.com> wrote:

I think it is good to add as it is also described in univocity library but
I am not too sure if it is worth exposing an option that has currently a
little bug. Maybe we could close for now and re-open this after we bump it
up to 2.4.0.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/17177#issuecomment-286382914>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AGxYPh_jy-HOk3Rk7YK_Ma1YO3gPHySTks5rlm3HgaJpZM4MUCg4>
.

jbax commented:
2.4.0 released, thank you guys!

Commit:
resolve conflicts

ep1804 commented:
I agree with you @HyukjinKwon , this PR will be closed for now and re-open.

And, thank you for the notice @jbox !

HyukjinKwon commented:
@ep1804 @jbax Thank you. I will cc and inform you both when I happen to see a PR bumping up the version to 2.4.0 (or probably I guess I will).
